File: petAI-backend/app/models/daily_activity.py
from __future__ import annotations

from datetime import datetime, timezone, date

from . import db


def _utcnow() -> datetime:
    return datetime.now(timezone.utc)


class DailyActivity(db.Model):
    __tablename__ = "daily_activities"

    id = db.Column(db.Integer, primary_key=True)
    user_id = db.Column(db.Integer, db.ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    interest_id = db.Column(db.Integer, db.ForeignKey("areas.id", ondelete="CASCADE"), nullable=False, index=True)
    activity_type_id = db.Column(db.Integer, db.ForeignKey("activity_types.id", ondelete="CASCADE"), nullable=False)
    goal_id = db.Column(db.Integer, db.ForeignKey("goals.id", ondelete="SET NULL"))
    title = db.Column(db.String(255), nullable=False)
    scheduled_for = db.Column(db.Date, nullable=False, index=True)
    todo_date = db.Column(db.Date, nullable=False, index=True)
    status = db.Column(db.String(32), default="pending", nullable=False)
    completed_at = db.Column(db.DateTime(timezone=True))
    xp_awarded = db.Column(db.Integer)
    created_at = db.Column(db.DateTime(timezone=True), default=_utcnow, nullable=False)
    updated_at = db.Column(db.DateTime(timezone=True), default=_utcnow, onupdate=_utcnow, nullable=False)

    interest = db.relationship("Area")
    activity_type = db.relationship("ActivityType")
    goal = db.relationship("Goal")

    __table_args__ = (
        db.UniqueConstraint(
            "user_id",
            "activity_type_id",
            "scheduled_for",
            "title",
            name="uq_user_activitytype_date_title",
        ),
    )

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "user_id": self.user_id,
            "interest_id": self.interest_id,
            "activity_type_id": self.activity_type_id,
            "goal_id": self.goal_id,
            "title": self.title,
            "scheduled_for": self.scheduled_for.isoformat(),
            "todo_date": self.todo_date.isoformat(),
            "status": self.status,
            "completed_at": self.completed_at.isoformat() if self.completed_at else None,
            "xp_awarded": self.xp_awarded,
        }

File: petAI-frontend/lib/models/daily_activity.dart
class DailyActivity {
  const DailyActivity({
    required this.id,
    required this.interestId,
    required this.activityTypeId,
    required this.title,
    required this.scheduledFor,
    required this.status,
    this.goalId,
    this.goalUnit,
    this.completedAt,
    this.xpAwarded,
  });

  factory DailyActivity.fromJson(Map<String, dynamic> json) {
    return DailyActivity(
      id: json['id'] as int? ?? 0,
      interestId: json['interest_id'] as int? ?? 0,
      activityTypeId: json['activity_type_id'] as int? ?? 0,
      title: json['title'] as String? ?? '',
      scheduledFor:
          DateTime.tryParse(json['scheduled_for'] as String? ?? '') ??
          DateTime.now(),
      status: json['status'] as String? ?? 'pending',
      goalId: json['goal_id'] as int?,
      goalUnit: (json['goal_unit'] as String?)?.trim(),
      completedAt: DateTime.tryParse(json['completed_at'] as String? ?? ''),
      xpAwarded: json['xp_awarded'] as int?,
    );
  }

  final int id;
  final int interestId;
  final int activityTypeId;
  final int? goalId;
  final String? goalUnit;
  final String title;
  final DateTime scheduledFor;
  final String status;
  final DateTime? completedAt;
  final int? xpAwarded;

  bool get isCompleted => status.toLowerCase() == 'completed';
}
